File: gawk-broadcast/internal/engine/stats.go
package engine

// Stats is what the engine can honestly see, mirroring the browser's
// BroadcastStats where the two overlap so a native diagnostics dump can be
// read next to a browser one (R9, docs/13).
//
// Decision 20 governs the gaps: report what we can see, mark what we cannot,
// fabricate nothing. The GStreamer child owns capture, so pre-encode capture
// fps is genuinely unavailable here — CaptureFpsAvailable is false and the
// shells print "n/a". Deriving it from the *requested* rate would be a
// plausible-looking number that answers the exact question ("is the source
// keeping up?") wrongly, which is worse than a gap.
// Every field carries an explicit JSON tag, and the tags are the browser's
// lowerCamelCase spelling rather than Go's exported names. That is load-bearing
// rather than cosmetic: this struct is serialized directly into an R28
// telemetry batch, and everything downstream — schema.BroadcasterFields, the
// rollup's curated series, diagnose()'s rules, the dip detector, the dashboard
// — matches on the browser's names. Untagged, the native broadcaster reported
// `EncoderFps` where every reader looked for `encoderFps`: stored faithfully,
// readable by nothing. See stats_test.go, which pins the key set.
//
// Fields the two producers genuinely share take the browser's name outright
// (the funnel, the D17 target trio, the audio lane). Fields only this engine
// has keep their own name in the same case (`encoder`, `capturePath`,
// `audioSource`) rather than being forced into a browser field whose values
// mean something else — one column, one vocabulary.
type Stats struct {
	// Encoder is the cascade candidate actually in use ("vulkanh264enc",
	// "nvh264enc", "vah264enc"), or "" before one is chosen.
	Encoder string `json:"encoder,omitempty"`
	// Codec is the SPS-derived codec string sent to viewers ("avc1.42E02A").
	Codec string `json:"codec,omitempty"`
	// Width/Height/Fps/BitrateBps are the session's fixed rung (Decision 9),
	// and they are what R28's D17 target rules read — hence the `target*`
	// names, which are the browser's for the same quantities.
	Width      int `json:"targetWidth,omitempty"`
	Height     int `json:"targetHeight,omitempty"`
	Fps        int `json:"targetFps,omitempty"`
	BitrateBps int `json:"targetBitrateBps,omitempty"`

	// CaptureFpsAvailable is always false: see Decision 20 above. The field
	// exists so the gap is explicit in the JSON rather than an absent key a
	// reader might mistake for zero.
	CaptureFpsAvailable bool `json:"captureFpsAvailable"`

	// CapturePath says how frames cross the pipewiresrc boundary:
	// "zero-copy" (DMA-BUF stays on the GPU) or "system-memory" (one CPU
	// copy per frame); "" before capture starts.
	CapturePath string `json:"capturePath,omitempty"`

	// EncodedFrames counts access units demuxed from the child.
	EncodedFrames uint64 `json:"encodedFrames"`
	Keyframes     uint64 `json:"keyframes"`

	// KeyframeIntervalMs is an EMA of the wall-clock spacing between
	// keyframes leaving the encoder, measured on AU arrival stamps.
	// GOPMs is the *target*; the encoders take a frame count derived from
	// the nominal fps, so damage-driven capture running under that rate
	// stretches the real cadence proportionally — this stat is what makes
	// that visible. Available is false until two keyframes have arrived.
	KeyframeIntervalAvailable bool    `json:"keyframeIntervalAvailable"`
	KeyframeIntervalMs        float64 `json:"keyframeIntervalMs"`

	// ParityLevel is how many R29 forward-parity symbols this producer emits
	// per delta frame (docs/34). It mirrors the browser's
	// BroadcastStats.parityLevel: the level the RELAY advertised, not a local
	// preference — a producer cannot choose to protect a stream the fleet has
	// not asked it to, because the relay is what filters symbols per
	// subscriber. 0 against a relay predating R29 or configured off.
	ParityLevel      int    `json:"parityLevel"`
	ParityChunksSent uint64 `json:"parityChunksSent"`
	ParityBytesSent  uint64 `json:"parityBytesSent"`
	// SentFrames counts frames whose bytes reached the transport without
	// error — R9's funnel stage 4, "actually sent".
	SentFrames uint64 `json:"sentFrames"`
	// EncoderFps is the rate access units leave the child, and SentFps the
	// rate they actually leave the machine. Both are windowed over the stats
	// interval (the browser computes them the same way). A gap between them
	// localizes the bottleneck to the uplink rather than the encoder.
	EncoderFps float64 `json:"encoderFps"`
	SentFps    float64 `json:"sentFps"`

	DatagramsSent uint64 `json:"datagramsSent"`
	BytesSent     uint64 `json:"bytesSent"`
	ConfigsSent   uint64 `json:"configsSent"`

	// Keyframe uni streams (R8 discipline, applied at the source by
	// Decision 12).
	KeyframeStreamsSent       uint64 `json:"keyframeStreamsSent"`
	KeyframeStreamsFailed     uint64 `json:"keyframeStreamsFailed"`
	KeyframeStreamsSuperseded uint64 `json:"keyframeStreamsSuperseded"`
	KeyframeBytesSent         uint64 `json:"keyframeBytesSent"`

	// FramesDroppedAtSend counts frames abandoned mid-chunking after a
	// datagram send failure (Decision 12): the remainder is dead weight the
	// viewer's reassembler would discard anyway.
	FramesDroppedAtSend uint64 `json:"framesDroppedAtSend"`

	// TimeSyncRttMs is the self-owned broadcaster↔relay RTT, and
	// TimeSyncOffsetUs the relay-clock offset published as ClockMapping.
	// Available is false until the first pong lands.
	TimeSyncAvailable bool    `json:"timeSyncAvailable"`
	TimeSyncRttMs     float64 `json:"timeSyncRttMs"`
	TimeSyncOffsetUs  int64   `json:"timeSyncOffsetUs"`

	// ViewerCount is the live "N watching" number the relay pushes (R18,
	// docs/23 — fleet-global in cluster mode). Available is false until the
	// first push lands (an old relay never sends one).
	ViewerCountAvailable bool   `json:"viewerCountAvailable"`
	ViewerCount          uint32 `json:"viewerCount"`

	// Resumes counts how many times auto-resume has reclaimed this broadcast
	// on a fresh relay session, and Resuming says whether a reclaim is in
	// flight right now (capture and encode still running, nothing reaching
	// viewers). A broadcast that quietly resumes every few minutes is a
	// working broadcast on a failing path, and only this counter says so.
	Resumes  uint64 `json:"resumes"`
	Resuming bool   `json:"resuming"`

	// System audio (R25, docs/28), mirroring the browser lane's audio stats
	// so the two dumps read alike.
	//
	// AudioState is the honest summary: "off" when the broadcaster asked for
	// silence, "unavailable" when audio was wanted and no source could give
	// it, "active" when packets are flowing, "error" when the bitstream
	// disagreed with the config we advertise (Decision 10). A machine with no
	// usable audio source is *not* an error, and this vocabulary says so.
	AudioState AudioState `json:"audioState,omitempty"`
	// AudioSource names the winning cascade candidate ("pipewire-monitor"),
	// which is the first thing to know when audio is present but wrong.
	AudioSource string `json:"audioSource,omitempty"`
	// The format actually advertised to viewers. Empty until a lane starts.
	AudioCodec      string `json:"audioCodec,omitempty"`
	AudioSampleRate int    `json:"audioSampleRate,omitempty"`
	AudioChannels   int    `json:"audioChannels,omitempty"`
	AudioBitrateBps int    `json:"audioBitrateBps,omitempty"`

	AudioPacketsSent uint64 `json:"audioPacketsSent"`
	AudioBytesSent   uint64 `json:"audioBytesSent"`
	AudioConfigsSent uint64 `json:"audioConfigsSent"`
	// AudioPacketsDropped counts packets lost at the send path — oversize, or
	// a datagram the transport refused. Deliberately its own counter: audio
	// never touches FramesDroppedAtSend, so a glance at the two separates a
	// saturated uplink from an audio-only problem.
	AudioPacketsDropped uint64 `json:"audioPacketsDropped"`
}

File: gawk-broadcast-windows/crates/engine/src/stats.rs
//! Session counters, mirroring the Go engine's `Stats` — and, like it,
//! serialized with the BROWSER's lowerCamelCase field names. That is
//! load-bearing rather than cosmetic: this struct feeds the diagnostics
//! JSON (WB6) and the R28 telemetry batches (WB7), and everything
//! downstream — the schema's broadcaster fields, the rollup's curated
//! series, the dashboard — matches on the browser's names. The Go engine
//! shipped the mistake once: untagged, it reported `EncoderFps` where every
//! reader looked for `encoderFps` — stored faithfully, readable by nothing.
//! The key-set test below is the pin.
//!
//! Availability flags are explicit fields (`…Available` + value), exactly
//! as the Go struct spells them: an absent number and a zero must stay
//! distinguishable to readers. One honest improvement over the Linux
//! engine: capture fps is REAL here — we own capture (docs/38 D12) — where
//! the Go engine's `captureFpsAvailable` is permanently false because the
//! GStreamer child owns it.

use serde::Serialize;

fn is_zero_u32(v: &u32) -> bool {
    *v == 0
}
fn is_zero_u8(v: &u8) -> bool {
    *v == 0
}

#[derive(Debug, Clone, Default, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Stats {
    /// The encoder actually in use (e.g. "NVIDIA H264 Encoder MFT"), ""
    /// before one is chosen.
    #[serde(skip_serializing_if = "String::is_empty")]
    pub encoder: String,
    /// The SPS-derived codec string sent to viewers ("avc1.42E02A").
    #[serde(skip_serializing_if = "String::is_empty")]
    pub codec: String,
    /// The session's fixed rung — the browser's `target*` names, which is
    /// what R28's target rules read.
    #[serde(rename = "targetWidth", skip_serializing_if = "is_zero_u32")]
    pub width: u32,
    #[serde(rename = "targetHeight", skip_serializing_if = "is_zero_u32")]
    pub height: u32,
    #[serde(rename = "targetFps", skip_serializing_if = "is_zero_u32")]
    pub fps: u32,
    #[serde(rename = "targetBitrateBps", skip_serializing_if = "is_zero_u32")]
    pub bitrate_bps: u32,

    pub capture_fps_available: bool,
    pub capture_fps: f64,
    /// How frames reach the encoder: "zero-copy" or "tone-mapped" (WB3).
    #[serde(skip_serializing_if = "String::is_empty")]
    pub capture_path: String,

    pub encoded_frames: u64,
    pub keyframes: u64,

    /// EMA (α = 0.3) of the wall-clock keyframe spacing — measured and
    /// displayed, never assumed. Available is false until two keyframes.
    pub keyframe_interval_available: bool,
    pub keyframe_interval_ms: f64,

    /// The level the RELAY advertised (never a local preference).
    pub parity_level: u8,
    pub parity_chunks_sent: u64,
    pub parity_bytes_sent: u64,

    pub sent_frames: u64,
    pub encoder_fps: f64,
    pub sent_fps: f64,

    pub datagrams_sent: u64,
    pub bytes_sent: u64,
    pub configs_sent: u64,

    pub keyframe_streams_sent: u64,
    pub keyframe_streams_failed: u64,
    pub keyframe_streams_superseded: u64,
    pub keyframe_bytes_sent: u64,

    pub frames_dropped_at_send: u64,

    pub time_sync_available: bool,
    pub time_sync_rtt_ms: f64,
    pub time_sync_offset_us: i64,

    pub viewer_count_available: bool,
    pub viewer_count: u32,

    pub resumes: u64,
    pub resuming: bool,

    /// "off" | "unavailable" | "active" | "error" — a machine with no
    /// usable audio source is NOT an error, and this vocabulary says so.
    #[serde(skip_serializing_if = "String::is_empty")]
    pub audio_state: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub audio_source: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub audio_codec: String,
    #[serde(skip_serializing_if = "is_zero_u32")]
    pub audio_sample_rate: u32,
    #[serde(skip_serializing_if = "is_zero_u8")]
    pub audio_channels: u8,
    #[serde(skip_serializing_if = "is_zero_u32")]
    pub audio_bitrate_bps: u32,

    pub audio_packets_sent: u64,
    pub audio_bytes_sent: u64,
    pub audio_configs_sent: u64,
    pub audio_packets_dropped: u64,
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::BTreeSet;

    // The key-set pin, mirroring the Go stats_test.go: every serialized key
    // is a browser name, spelled exactly. A drift here means telemetry that
    // is stored faithfully and readable by nothing.
    #[test]
    fn serialized_keys_are_the_browser_names() {
        let full = Stats {
            encoder: "e".into(),
            codec: "avc1.42E02A".into(),
            width: 1920,
            height: 1080,
            fps: 60,
            bitrate_bps: 16_000_000,
            capture_path: "zero-copy".into(),
            audio_state: "active".into(),
            audio_source: "process-loopback".into(),
            audio_codec: "opus".into(),
            audio_sample_rate: 48_000,
            audio_channels: 2,
            audio_bitrate_bps: 128_000,
            ..Default::default()
        };
        let v = serde_json::to_value(&full).unwrap();
        let keys: BTreeSet<String> = v.as_object().unwrap().keys().cloned().collect();
        let want: BTreeSet<String> = [
            "encoder",
            "codec",
            "targetWidth",
            "targetHeight",
            "targetFps",
            "targetBitrateBps",
            "captureFpsAvailable",
            "captureFps",
            "capturePath",
            "encodedFrames",
            "keyframes",
            "keyframeIntervalAvailable",
            "keyframeIntervalMs",
            "parityLevel",
            "parityChunksSent",
            "parityBytesSent",
            "sentFrames",
            "encoderFps",
            "sentFps",
            "datagramsSent",
            "bytesSent",
            "configsSent",
            "keyframeStreamsSent",
            "keyframeStreamsFailed",
            "keyframeStreamsSuperseded",
            "keyframeBytesSent",
            "framesDroppedAtSend",
            "timeSyncAvailable",
            "timeSyncRttMs",
            "timeSyncOffsetUs",
            "viewerCountAvailable",
            "viewerCount",
            "resumes",
            "resuming",
            "audioState",
            "audioSource",
            "audioCodec",
            "audioSampleRate",
            "audioChannels",
            "audioBitrateBps",
            "audioPacketsSent",
            "audioBytesSent",
            "audioConfigsSent",
            "audioPacketsDropped",
        ]
        .into_iter()
        .map(String::from)
        .collect();
        assert_eq!(keys, want);
    }

    // Empty-vs-zero honesty: optional strings and the rung vanish when
    // unset; availability-paired numbers stay present with their flag.
    #[test]
    fn unset_optionals_are_absent_not_zero() {
        let v = serde_json::to_value(Stats::default()).unwrap();
        let obj = v.as_object().unwrap();
        for absent in [
            "encoder",
            "codec",
            "targetWidth",
            "audioState",
            "audioCodec",
        ] {
            assert!(!obj.contains_key(absent), "{absent} should be absent");
        }
        assert_eq!(obj["captureFpsAvailable"], false);
        assert_eq!(obj["keyframeIntervalAvailable"], false);
        assert_eq!(obj["viewerCountAvailable"], false);
        assert_eq!(obj["timeSyncAvailable"], false);
    }
}
